### ISSUE TITLE
TTNN Advanced Indexing ops support for VAD-v2 model

### ISSUE THREAD

**Issue by sdawle-TT** [tenstorrent/tt-metal#32734]
# TTNN Advanced and Boolean Indexing Ops support for VAD v2 model

---

## Current Situation

The VADv2 (Vectorized Autonomous Driving v2) model requires advanced and boolean indexing ops support.Currently, this is implemented using pytorch **tensor[range(n), indices]** and operation

1. Converting TTNN tensors to PyTorch tensors
2. Performing boolean masking on CPU/PyTorch
3. Converting back to TTNN tensors

This creates a **significant performance bottleneck** and occurs per-batch-item.

---

### Usage in VADv2 Model

**1. **Map Query Selection**:  - Advanced Indexing**

**Actual Code:
tt_head.py - Lines (703-707)**

       **# CONVERT TO TORCH**
        min_map_pos = ttnn.to_torch(min_map_pos)
        min_map_pos_idx = ttnn.to_torch(min_map_pos_idx)

        **# ADVANCED INDEXING OPERATION**
        min_map_pos = min_map_pos[range(min_map_pos.shape[0]), min_map_pos_idx]  # [B*P, 2]

        **# CONVERT BACK FROM TORCH**
        min_map_pos = ttnn.from_torch(min_map_pos, dtype=ttnn.bfloat16, device=self.dev

**Explanation:**
min_map_pos[range(min_map_pos.shape[0]), min_map_pos_idx] uses range(n) for rows (selecting all rows in order) and min_map_pos_idx for columns (selecting different columns per row).

Your Operation Explained
If min_map_pos has shape [BxP, C] and min_map_pos_idx has shape [BxP], then:

Row 0: selects min_map_pos[0, min_map_pos_idx[0]]
Row 1: selects min_map_pos[1, min_map_pos_idx[1]]
Row i: selects min_map_pos[i, min_map_pos_idx[i]]
The result is a 1D tensor of shape [BxP] containing one selected element per row.

**Simplified Example:**
```
import torch  
  
# Example matching your use case  
min_map_pos = torch.tensor([[1.0, 2.0, 3.0],  
                            [4.0, 5.0, 6.0],  
                            [7.0, 8.0, 9.0]])  # Shape: [3, 3]  
  
min_map_pos_idx = torch.tensor([2, 0, 1])  # Shape: [3]  
  
# Advanced indexing  
result = min_map_pos[range(3), min_map_pos_idx]  
# Result: tensor([3.0, 4.0, 8.0])  
# Explanation: [row 0 col 2, row 1 col 0, row 2 col 1]
```

**2. **Motion Query Selection** - Boolean Indexing**

**Actual Code:
tt_head.py - Lines (995 - 1015)**

```
for i in range(query_score.shape[0]):
            dim = query.shape[-1]
            valid_qnum = ttnn.sum(query_idx[i])
            valid_qnum_int = int(valid_qnum.item()) if hasattr(valid_qnum, "item") else int(valid_qnum)

            **# FORCED CONVERSION TO TORCH**
            query = ttnn.to_torch(query)
            query_pos = ttnn.to_torch(query_pos)
            query_idx = ttnn.to_torch(query_idx, dtype=torch.bool)

            **# BOOLEAN INDEXING OPERATION**
            valid_query = query[i, query_idx[i]]
            valid_query_pos = query_pos[i, query_idx[i]]

            pad_qnum = batch_max_qnum_int - valid_qnum_int
            padding_mask = torch.tensor([False], device="cpu").repeat(batch_max_qnum_int)
            if pad_qnum != 0:
                valid_query = torch.cat([valid_query, torch.zeros((pad_qnum, dim), device=query_score.device)], dim=0)
                valid_query_pos = torch.cat(
                    [valid_query_pos, torch.zeros((pad_qnum, 2), device=query_score.device)], dim=0
                )
                padding_mask[valid_qnum_int:] = True

            **# CONVERTING BACK TO TTNN**
            valid_query = ttnn.from_torch(valid_query, dtype=ttnn.bfloat16, device=self.device)
            valid_query_pos = ttnn.from_torch(valid_query_pos, dtype=ttnn.bfloat16, device=self.device)
            padding_mask = ttnn.from_torch(padding_mask, dtype=ttnn.bfloat16, device=self.device)
```

**Explanation**
This code uses a boolean mask to filter elements from tensors:

valid_query = query[i, query_idx[i]]  
valid_query_pos = query_pos[i, query_idx[i]]
Here, query_idx[i] is a boolean tensor (converted via dtype=torch.bool) that acts as a mask. For a 2D tensor like query[i], the boolean mask selects only the elements where the mask is True.

**Simplified Example**
```
import torch  
  
# Create a 2D tensor  
query = torch.tensor([[1.0, 2.0, 3.0, 4.0],  
                      [5.0, 6.0, 7.0, 8.0]])  
  
# Boolean mask for row 0  
query_idx = torch.tensor([True, False, True, False])  
  
# Boolean indexing  
result = query[0, query_idx]  
# Result: tensor([1.0, 3.0])  # Only elements at positions 0 and 2
```
```


### Please complete the following environment information

OS: Ubuntu 22.04
Device: N300

**Comment by cmaryanTT** [comment]
It looks like this can be lowered to some existing TMs. Here's Claude's approach in torch:
import torch


```
import torch


def advanced_indexing(tensor, indices):
    """
    Implements tensor[range(n), indices] using gather.

    Args:
        tensor: [N, C] input tensor
        indices: [N] integer indices
    Returns:
        [N] tensor with selected elements
    """
    N = tensor.shape[0]
    # Reshape indices from [N] to [N, 1] for gather
    indices_expanded = indices.reshape(N, 1)
    result = torch.gather(tensor, dim=1, index=indices_expanded)
    # Reshape result from [N, 1] to [N]
    return result.reshape(N)


def boolean_indexing(tensor, mask):
    """
    Implements tensor[mask] using gather.

    Args:
        tensor: [N] or [N, C] input tensor
        mask: [N] boolean mask
    Returns:
        [M] or [M, C] tensor where M = mask.sum()
    """
    # Convert boolean mask to integer indices
    indices = torch.nonzero(mask, as_tuple=True)[0]
    M = indices.shape[0]

    if tensor.dim() == 1:
        return torch.gather(tensor, dim=0, index=indices)
    else:
        # For 2D, repeat indices to match all columns
        C = tensor.shape[-1]
        # Reshape [M] to [M, 1], then repeat along dim 1 to get [M, C]
        indices_expanded = indices.reshape(M, 1).repeat(1, C)
        return torch.gather(tensor, dim=0, index=indices_expanded)


# Test 1: Advanced indexing
print("=== Test 1: Advanced Indexing ===")
min_map_pos = torch.tensor([[1.0, 2.0, 3.0],
                            [4.0, 5.0, 6.0],
                            [7.0, 8.0, 9.0]])
min_map_pos_idx = torch.tensor([2, 0, 1])

# Original operation
expected = min_map_pos[range(3), min_map_pos_idx]
print(f"Expected: {expected}")

# Our implementation
result = advanced_indexing(min_map_pos, min_map_pos_idx)
print(f"Result:   {result}")
print(f"Match: {torch.allclose(expected, result)}")

# Test 2: Boolean indexing
print("\n=== Test 2: Boolean Indexing ===")
query = torch.tensor([[1.0, 2.0, 3.0, 4.0],
                      [5.0, 6.0, 7.0, 8.0]])
query_idx = torch.tensor([True, False, True, False])

# Original operation
expected = query[0, query_idx]
print(f"Expected: {expected}")

# Our implementation
result = boolean_indexing(query[0], query_idx)
print(f"Result:   {result}")
print(f"Match: {torch.allclose(expected, result)}")


```

**Comment by cmaryanTT** [comment]
@sdawle-TT - can you please try out the approach above in the model, porting the code above to TTNN.

**Comment by sdawle-TT** [comment]
Thanks @cmaryanTT, I will try the above approach on TTNN and report back

**Comment by sdawle-TT** [comment]
@cmaryanTT 
cc: @dvartaniansTT 
I have implemented the advanced indexing and boolean indexing functions here in [common.py](https://github.com/tenstorrent/tt-metal/blob/sdawle/vadv2_demo_v2/models/experimental/vadv2/tt/common.py)
I have been able to use the advanced_indexing functions in [tt_head.py](https://github.com/tenstorrent/tt-metal/blob/sdawle/vadv2_demo_v2/models/experimental/vadv2/tt/tt_head.py) module in vadv2.

The boolean_indexing implementation not used in tt_head due to implementation challenges resulting in worse performance than torch implementation

Overall adding advanced indexing did not improve the overall perf but we have been able to move the torch implementation to ttnn.

**Comment by cmaryanTT** [comment]
Revisit perf in January. @sdawle-TT - please post a perf breakdown before we continue looking at perf.

**Comment by fahmedchanTT** [comment]
@sdawle-TT Need the sanity check, Ops side will look at in Jan. (repeat from above)

**Comment by sdawle-TT** [comment]
Thanks @fahmedchanTT for the update, we can revisit this in Jan
